[WebGPU+Swift] Factor some unsafe code into StdLibExtras.swift

geoffreygaren opened the pull request:
#### 59fd40b9b902e6c949ecefa27c7f832b9c8c92b2
<pre>
[WebGPU+Swift] Factor some unsafe code into StdLibExtras.swift
<a href="https://bugs.webkit.org/show_bug.cgi?id=302024">https://bugs.webkit.org/show_bug.cgi?id=302024</a>
<a href="https://rdar.apple.com/164104484">rdar://164104484</a>

Reviewed by Richard Robinson and Mike Wyrzykowski.

These operations are safe in fact, even though they use &apos;unsafe&apos; in their
implementations.

Factoring them into a separate file helps clarify which pieces of WebGPU still
have safety work left to do.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.swift:
(WebGPU.copy(from:offset:)): It&apos;s now a bit clearer that the memory copy is in
fact safe; it&apos;s the lifetime of the buffer that&apos;s unsafe.

* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.errorValidatingCopyTextureToTexture(_:destination:copySize:)): It&apos;s now
a bit clearer that this code is perfectly safe.

* Source/WebGPU/WebGPU/Queue.swift:
(Queue_writeBuffer_thunk(_:buffer:bufferOffset:data:)):
(WebGPU.writeBuffer(_:bufferOffset:data:)):
* Source/WebGPU/WebGPU/StdLibExtras.swift: Added. Holding tank for functionality
that has agreement in principle to be added to Swift or the Swift standard library,
but hasn&apos;t been added yet.
(MutableSpan.copyMemory(from:)):

Canonical link: <a href="https://commits.webkit.org/302614@main">https://commits.webkit.org/302614@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e92e04e1d7b0695a0925c82ebdb2161002c6761f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129692 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1953 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40548 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137080 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/81152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/287444fd-57b4-4745-ab7d-110f743ad866) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1842 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98796 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/81152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ea70081b-8a9c-475c-8d1b-617d9ef5a273) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132639 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/116154 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79470 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e792298b-99f2-4c2d-8639-5cb375b111e2) 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80352 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34785 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139562 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1747 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1639 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107308 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1792 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112498 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107175 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1408 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31002 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54497 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20233 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1820 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65189 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1634 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/1669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1741 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

adetaylor commented on Source/WebGPU/WebGPU/StdLibExtras.swift:
```diff
@@ -0,0 +1,46 @@
+
+// FIXME(rdar://130765784): We should be able use the built-in === for AnyObject
+func === (_ a: WGPUTexture, _ b: WGPUTexture) -> Bool {
+    unsafe unsafeBitCast(a, to: UnsafeRawPointer.self) == unsafeBitCast(b, to: UnsafeRawPointer.self)
```
```suggestion
    // Safety: references to C++ types are currently guaranteed to be represented as raw pointers behind the scenes
    unsafe unsafeBitCast(a, to: UnsafeRawPointer.self) == unsafeBitCast(b, to: UnsafeRawPointer.self)
```
I assume this is true?

adetaylor commented on Source/WebGPU/WebGPU/StdLibExtras.swift:
```diff
@@ -0,0 +1,46 @@
+extension MutableSpan where Element: BitwiseCopyable {
+    @_lifetime(self: copy self)
+    mutating func copyMemory(from source: Span<Element>) {
+        // This is lifetime safe because we have exclusive access to 'self' and we don't escape 'selfBuffer'
```
Nit: I'd love it if each of these comment lines is prefixed with the word "Safety:" just to help set a consistent pattern across the rest of WebKit that such safety comments are required.
(These are great by the way, thanks!)

adetaylor reviewed (approved):
LGTM modulo the fact that I can't find docs for `_consumingExtracting`. Thanks for the safety comments!

rr-codes reviewed (changes requested):
.

rr-codes commented on Source/WebGPU/WebGPU/StdLibExtras.swift:
```diff
@@ -0,0 +1,47 @@
+            unsafe source.withUnsafeBufferPointer { sourceBuffer in
+                // This is bounds safe because we do a manual bounds check
+                // This is type safe because we statically declare that our element types match and are BitwiseCopyable
+                precondition(sourceBuffer.count <= selfBuffer.count)
```
(this is a release assert, just checking to make sure that’s expected)

rr-codes commented on Source/WebGPU/WebGPU/StdLibExtras.swift:
```diff
@@ -0,0 +1,47 @@
+    @_lifetime(self: copy self)
+    mutating func copyMemory(from source: Span<Element>) {
+        // This is lifetime safe because we have exclusive access to 'self' and we don't escape 'selfBuffer'
+        unsafe self.withUnsafeMutableBufferPointer { selfBuffer in
```
`self.` is not needed and should be removed

rr-codes commented on Source/WebGPU/WebGPU/StdLibExtras.swift:
```diff
@@ -0,0 +1,47 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
```
this should use // comments like other swift file copyrights

mwyrzykowski replied:
Just wondering, why should Swift use different style from the rest of WebKit? Isn't that inconsistent within WebKit?

geoffreygaren replied:
@rr-codes and I chatted and I'm going to leave this as-is, and post a follow-up to use the same license text style in our .swift files.

mwyrzykowski replied:
Sounds good 👍 

At some point someone probably should start a discussion on webkit-dev regarding Swift coding style, whether it should match the rest of WebKit or match other Swift frameworks. It’s not really documented on https://webkit.org/code-style-guidelines/

rr-codes commented on Source/WebGPU/WebGPU/Buffer.swift:
```diff
@@ -26,13 +26,12 @@
-        }
+    func copy(from source: Span<UInt8>, offset: Int) {
+        // FIXME (rdar://161274084): Swift doesn't have a lifetime-safe way to return a borrowed value from a refcounted object yet.
+        let bufferContents = unsafe MutableSpan<UInt8>(_unsafeCxxSpan: getBufferContents())
```
```suggestion
        let bufferContents = unsafe MutableSpan(_unsafeCxxSpan: getBufferContents())
```

this _should_ work

rr-codes commented on Source/WebGPU/WebGPU/StdLibExtras.swift:
```diff
@@ -0,0 +1,47 @@
+}
+
+// FIXME(rdar://130765784): We should be able use the built-in === for AnyObject
+func === (_ a: WGPUTexture, _ b: WGPUTexture) -> Bool {
```
```suggestion
func === (_ lhs: WGPUTexture, _ rhs: WGPUTexture) -> Bool {
```

this is the usual variable naming convention for these sorts of functions

geoffreygaren committed:
[WebGPU+Swift] Factor some unsafe code into StdLibExtras.swift
https://bugs.webkit.org/show_bug.cgi?id=302024
rdar://164104484

Reviewed by Richard Robinson and Mike Wyrzykowski.

These operations are safe in fact, even though they use 'unsafe' in their
implementations.

Factoring them into a separate file helps clarify which pieces of WebGPU still
have safety work left to do.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.swift:
(WebGPU.copy(from:offset:)): It's now a bit clearer that the memory copy is in
fact safe; it's the lifetime of the buffer that's unsafe.

* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.errorValidatingCopyTextureToTexture(_:destination:copySize:)): It's now
a bit clearer that this code is perfectly safe.

* Source/WebGPU/WebGPU/Queue.swift:
(Queue_writeBuffer_thunk(_:buffer:bufferOffset:data:)):
(WebGPU.writeBuffer(_:bufferOffset:data:)):
* Source/WebGPU/WebGPU/StdLibExtras.swift: Added. Holding tank for functionality
that has agreement in principle to be added to Swift or the Swift standard library,
but hasn't been added yet.
(MutableSpan.copyMemory(from:)):

Canonical link: https://commits.webkit.org/302614@main
